ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/committer/transaction/validation/ModelConformanceValidator.scala

```diff
@@ -170,37 +182,67 @@ private[transaction] class ModelConformanceValidator(engine: Engine, metrics: Me
       } yield pkg
     }
 
-  private def lookupKey(
-      commitContext: CommitContext,
-      knownKeys: Map[DamlContractKey, Value.ContractId],
+  private[validation] def lookupKey(
+      contractKeyInputs: Map[GlobalKey, KeyInput]
   )(key: GlobalKeyWithMaintainers): Option[Value.ContractId] = {
-    // we don't check whether the contract is active or not, because in we might not have loaded it earlier.
-    // this is not a problem, because:
-    // a) if the lookup was negative and we actually found a contract,
-    //    the transaction validation will fail.
-    // b) if the lookup was positive and its result is a different contract,
-    //    the transaction validation will fail.
-    // c) if the lookup was positive and its result is the same contract,
-    //    - the authorization check ensures that the submitter is in fact allowed
-    //      to lookup the contract
-    //    - the separate contract keys check ensures that all contracts pointed to by
-    //    contract keys respect causal monotonicity.
-    val stateKey = Conversions.globalKeyToStateKey(key.globalKey)
-    val contractId = for {
-      stateValue <- commitContext.read(stateKey)
-      if stateValue.getContractKeyState.getContractId.nonEmpty
-    } yield decodeContractId(stateValue.getContractKeyState.getContractId)
+    contractKeyInputs.get(key.globalKey) match {
+      case Some(KeyActive(cid)) => Some(cid)
+      case _ => None
+    }
+  }
+
+  private[validation] def validateCausalMonotonicity(
+      transactionEntry: DamlTransactionEntrySummary,
+      commitContext: CommitContext,
+      rejections: Rejections,
+  )(implicit loggingContext: LoggingContext): StepResult[DamlTransactionEntrySummary] = {
 
-    // If the key was not in state inputs, then we look whether any of the accessed contracts has
-    // the key we're looking for. This happens with "fetchByKey" where the key lookup is not
-    // evidenced in the transaction. The activeness of the contract is checked when it is fetched.
-    contractId.orElse {
-      knownKeys.get(stateKey.getContractKey)
+    val inputContracts: Map[Value.ContractId, DamlContractState] = commitContext
+      .collectInputs {
+        case (key, Some(value)) if value.hasContractState =>
+          Conversions.stateKeyToContractId(key) -> value.getContractState
+      }
+
+    val isCasualMonotonicityHeld = transactionEntry.transaction.inputContracts.forall {
+      contractId =>
+        val inputContractState = inputContracts(contractId)
+        val activeAt = Option(inputContractState.getActiveAt).map(parseTimestamp)
+        activeAt.exists(transactionEntry.ledgerEffectiveTime >= _)
     }
+
+    if (isCasualMonotonicityHeld)
+      StepContinue(transactionEntry)
+    else
+      rejections.buildRejectionStep(
+        transactionEntry,
+        RejectionReasonV0.InvalidLedgerTime("Causal monotonicity violated"),
```

Comment:
@gerolf-da @moritzkiefer-da @miklos-da The existing tests haven't detected any changes in rejection reasons, as well as error codes. I would expect there is a change, at least here. Even though the test coverage is undoubtedly greater than before this PR, it would be good to write more tests. Also, I have no idea how it behaves in racy conditions like in the recent incident. Do you think it would be worth investing some time to check it (possibly as a separate story)?